client/rest-high-level/src/test/java/org/elasticsearch/client/SecurityIT.java

```diff
@@ -0,0 +1,108 @@
+ */
+
+package org.elasticsearch.client;
+
+import org.apache.http.client.methods.HttpDelete;
+import org.elasticsearch.ElasticsearchStatusException;
+import org.elasticsearch.client.security.AuthenticateResponse;
+import org.elasticsearch.client.security.PutUserRequest;
+import org.elasticsearch.client.security.PutUserResponse;
+import org.elasticsearch.client.security.RefreshPolicy;
+import org.elasticsearch.common.CharArrays;
+
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.contains;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.empty;
+
+public class SecurityIT extends ESRestHighLevelClientTestCase {
+
+    public void testAuthenticate() throws Exception {
+        final SecurityClient securityClient = highLevelClient().security();
+        // test fixture: put enabled user
+        final PutUserRequest putUserRequest = randomPutUserRequest(true);
+        final PutUserResponse putUserResponse = execute(putUserRequest, securityClient::putUser, securityClient::putUserAsync);
+        assertThat(putUserResponse.isCreated(), is(true));
+
+        // authenticate correctly
+        final String basicAuthHeader = basicAuthHeader(putUserRequest.getUsername(), putUserRequest.getPassword());
+        final AuthenticateResponse authenticateResponse = execute(securityClient::authenticate, securityClient::authenticateAsync,
+                authorizationRequestOptions(basicAuthHeader));
+
+        assertThat(authenticateResponse.getUser().username(), is(putUserRequest.getUsername()));
+        if (putUserRequest.getRoles().isEmpty()) {
+            assertThat(authenticateResponse.getUser().roles(), is(empty()));
+        } else {
+            assertThat(authenticateResponse.getUser().roles(), contains(putUserRequest.getRoles().toArray()));
+        }
+        assertThat(authenticateResponse.getUser().metadata(), is(putUserRequest.getMetadata()));
+        assertThat(authenticateResponse.getUser().fullName(), is(putUserRequest.getFullName()));
+        assertThat(authenticateResponse.getUser().email(), is(putUserRequest.getEmail()));
+        assertThat(authenticateResponse.enabled(), is(true));
+
+        // delete user
+        final Request deleteUserRequest = new Request(HttpDelete.METHOD_NAME, "/_xpack/security/user/" + putUserRequest.getUsername());
+        highLevelClient().getLowLevelClient().performRequest(deleteUserRequest);
+
+        // authentication no longer works
+        ElasticsearchStatusException e = expectThrows(ElasticsearchStatusException.class, () -> execute(securityClient::authenticate,
+                securityClient::authenticateAsync, authorizationRequestOptions(basicAuthHeader)));
+        assertThat(e.getMessage(), containsString("unable to authenticate user [" + putUserRequest.getUsername() + "]"));
+    }
```

Comment:
Should we be adding authentication tests for when the user is missing, password is wrong, user is disabled . I feel like I know the answer, but I am still asking :)

Reply:
u know the answer ;)